File: Calculator/ios-native/CalculatorHub/Calculators/NewMathCalculators.swift
import Foundation

// NEW MATH CALCULATORS

// Logarithm Calculator
struct LogarithmResult { let result: Double; let formula: String }
class LogarithmCalculator {
    func naturalLog(_ value: Double) -> LogarithmResult { LogarithmResult(result: log(value), formula: "ln(\(value))") }
    func commonLog(_ value: Double) -> LogarithmResult { LogarithmResult(result: log10(value), formula: "log₁₀(\(value))") }
    func binaryLog(_ value: Double) -> LogarithmResult { LogarithmResult(result: log2(value), formula: "log₂(\(value))") }
    func customLog(_ value: Double, base: Double) -> LogarithmResult { LogarithmResult(result: log(value) / log(base), formula: "log\(Int(base))(\(value))") }
}

// Statistics Calculator
struct StatisticsResult { let mean, median: Double; let mode: [Double]?; let range: Double; let count: Int }
class StatisticsCalculator {
    func calculate(_ numbers: [Double]) -> StatisticsResult {
        guard !numbers.isEmpty else { return StatisticsResult(mean: 0, median: 0, mode: nil, range: 0, count: 0) }
        let sorted = numbers.sorted(); let count = numbers.count
        let mean = numbers.reduce(0, +) / Double(count)
        let median = count % 2 == 0 ? (sorted[count/2 - 1] + sorted[count/2]) / 2 : sorted[count/2]
        var freq: [Double: Int] = [:]; numbers.forEach { freq[$0, default: 0] += 1 }
        let maxFreq = freq.values.max() ?? 0
        let mode = maxFreq > 1 ? freq.filter { $0.value == maxFreq }.keys.sorted() : nil
        return StatisticsResult(mean: mean, median: median, mode: mode, range: sorted.last! - sorted.first!, count: count)
    }
}

// Summation Calculator
struct SummationResult { let sum, nthTerm: Double; let terms: [Double] }
class SummationCalculator {
    func arithmetic(firstTerm: Double, commonDiff: Double, numTerms: Int) -> SummationResult {
        let nthTerm = firstTerm + Double(numTerms - 1) * commonDiff
        let sum = Double(numTerms) / 2.0 * (2 * firstTerm + Double(numTerms - 1) * commonDiff)
        let terms = (0..<min(numTerms, 10)).map { firstTerm + Double($0) * commonDiff }
        return SummationResult(sum: sum, nthTerm: nthTerm, terms: terms)
    }
    func geometric(firstTerm: Double, commonRatio: Double, numTerms: Int) -> SummationResult {
        let nthTerm = firstTerm * pow(commonRatio, Double(numTerms - 1))
        let sum = commonRatio == 1 ? firstTerm * Double(numTerms) : firstTerm * (1 - pow(commonRatio, Double(numTerms))) / (1 - commonRatio)
        let terms = (0..<min(numTerms, 10)).map { firstTerm * pow(commonRatio, Double($0)) }
        return SummationResult(sum: sum, nthTerm: nthTerm, terms: terms)
    }
}

File: Calculator/android-native/app/src/main/java/com/qaantum/calculatorhub/calculators/NewMathCalculators.kt
package com.qaantum.calculatorhub.calculators

import kotlin.math.ln
import kotlin.math.log10
import kotlin.math.log2

// NEW MATH CALCULATORS

// Logarithm Calculator
data class LogarithmResult(val result: Double, val formula: String)
class LogarithmCalculator {
    fun naturalLog(value: Double) = LogarithmResult(ln(value), "ln($value)")
    fun commonLog(value: Double) = LogarithmResult(log10(value), "log₁₀($value)")
    fun binaryLog(value: Double) = LogarithmResult(log2(value), "log₂($value)")
    fun customLog(value: Double, base: Double) = LogarithmResult(ln(value) / ln(base), "log$base($value)")
}

// Statistics Calculator
data class StatisticsResult(val mean: Double, val median: Double, val mode: List<Double>?, val range: Double, val count: Int)
class StatisticsCalculator {
    fun calculate(numbers: List<Double>): StatisticsResult {
        if (numbers.isEmpty()) return StatisticsResult(0.0, 0.0, null, 0.0, 0)
        val sorted = numbers.sorted()
        val count = numbers.size
        val mean = numbers.sum() / count
        val median = if (count % 2 == 0) (sorted[count / 2 - 1] + sorted[count / 2]) / 2 else sorted[count / 2]
        val frequency = numbers.groupingBy { it }.eachCount()
        val maxFreq = frequency.values.maxOrNull() ?: 0
        val mode = if (maxFreq > 1) frequency.filter { it.value == maxFreq }.keys.toList() else null
        return StatisticsResult(mean, median, mode, range = sorted.last() - sorted.first(), count)
    }
}

// Summation Calculator
data class SummationResult(val sum: Double, val nthTerm: Double, val terms: List<Double>)
class SummationCalculator {
    fun arithmetic(firstTerm: Double, commonDiff: Double, numTerms: Int): SummationResult {
        val nthTerm = firstTerm + (numTerms - 1) * commonDiff
        val sum = numTerms / 2.0 * (2 * firstTerm + (numTerms - 1) * commonDiff)
        val terms = (0 until minOf(numTerms, 10)).map { i -> firstTerm + i * commonDiff }
        return SummationResult(sum, nthTerm, terms)
    }
    
    fun geometric(firstTerm: Double, commonRatio: Double, numTerms: Int): SummationResult {
        val nthTerm = firstTerm * Math.pow(commonRatio, (numTerms - 1).toDouble())
        val sum = if (commonRatio == 1.0) {
            firstTerm * numTerms
        } else {
            firstTerm * (1 - Math.pow(commonRatio, numTerms.toDouble())) / (1 - commonRatio)
        }
        val terms = (0 until minOf(numTerms, 10)).map { i -> firstTerm * Math.pow(commonRatio, i.toDouble()) }
        return SummationResult(sum, nthTerm, terms)
    }
}
